table/merging_iterator.cc

```diff
@@ -95,49 +289,64 @@ class MergingIterator : public InternalIterator {
-      child.SeekToLast();
+      child.iter.SeekToLast();
       AddToMaxHeapOrCheckStatus(&child);
     }
+
+    for (size_t i = 0; i < range_tombstone_iters_.size(); ++i) {
+      if (range_tombstone_iters_[i]) {
+        range_tombstone_iters_[i]->SeekToLast();
+        if (range_tombstone_iters_[i]->Valid()) {
+          // It is possible to be invalid due to snapshots.
+          InsertRangeTombstoneToMaxHeap(i);
+        }
+      }
+    }
+    FindPrevVisibleKey();
     direction_ = kReverse;
     current_ = CurrentReverse();
   }
 
+  // Position this merging iterator at the first key >= target (internal key).
+  // If range tombstones are present, keys covered by range tombstones are
+  // skipped, and this merging iter points to the first non-range-deleted key >=
+  // target after Seek(). If !Valid() and status().ok() then end of the iterator
+  // is reached.
+  //
+  // Internally, this involves positioning all child iterators at the first key
+  // >= target. If range tombstones are present, we apply a similar
+  // optimization, cascading seek, as in Pebble
+  // (https://github.com/cockroachdb/pebble). Specifically, if there is a range
+  // tombstone [start, end) that covers the target user key at level L, then
+  // this range tombstone must cover the range [target key, end) in all levels >
+  // L. So for all levels > L, we can pretend the target key is `end`. This
+  // optimization is applied at each level and hence the name "cascading seek".
+  // After a round of (cascading) seeks, the top of the heap is checked to see
+  // if it is covered by a range tombstone (see FindNextVisibleKey() for more
+  // detail), and advanced if so. The process is repeated until a
+  // non-range-deleted key is at the top of the heap, or heap becomes empty.
+  //
+  // As mentioned in comments above HeapItem, to make the checking of whether
+  // top of the heap is covered by some range tombstone efficient, we treat each
+  // range deletion [start, end) as two point keys and insert them into the same
+  // min/maxHeap_ where point iterators are. The set `active_` tracks the levels
+  // that have active range tombstones. If level L is in `active_`, and the
+  // point key at top of the heap is from level >= L, then the point key is
+  // within the internal key range of the range tombstone that
+  // range_tombstone_iters_[L] currently points to. For correctness reasoning,
+  // one invariant that Seek() guarantees is as follows. After Seek(), suppose
+  // `k` is the current key of level L's point iterator. Then for each range
+  // tombstone iterator at level <= L, it is at or before the first range
+  // tombstone with end key > `k`. This ensures that when level L's point
+  // iterator reaches top of the heap, `active_` is calculated correctly (it
+  // contains the covering range tombstone's level if there is one), since no
+  // range tombstone iterator was skipped beyond that point iterator's current
+  // key during Seek().
```

Comment:
> After Seek(), ...

Is this invariant really specific to Seek() or should it hold after every public scan/seek operation completes?

Reply:
Yes, good point, that is exactly true. Next()/Prev() have a stronger version of the invariant: all range tombstone iter above current_key is positioned **at** (instead of at or before) the first range tombstone that covers or after current_key, but this invariant is still true. 

Updated.